File: sdks/python/agentbox/sandbox.py
from __future__ import annotations

import base64
import json
from typing import AsyncIterator

import websockets

from .client import AgentBoxClient
from .types import ExecResult, FileEntry, PortForwardInfo, SandboxInfo


class Sandbox:
    """A sandboxed environment for executing code."""

    def __init__(self, id: str, client: AgentBoxClient):
        self.id = id
        self._client = client

    @classmethod
    def create(
        cls,
        url: str | None = None,
        memory_mb: int = 2048,
        vcpus: int = 2,
        network: bool = False,
        timeout: int = 3600,
        api_key: str | None = None,
    ) -> Sandbox:
        """Create a new sandbox. Boots a microVM in <300ms."""
        client = AgentBoxClient(url, api_key=api_key)
        data = client.post(
            "/sandboxes",
            json={
                "memory_mb": memory_mb,
                "vcpus": vcpus,
                "network": network,
                "timeout": timeout,
            },
        )
        return cls(id=data["id"], client=client)

    # ── Command execution ───────────────────────────────────────

    def exec(self, command: str, timeout: int = 30) -> ExecResult:
        """Execute a command and wait for completion."""
        data = self._client.post(
            f"/sandboxes/{self.id}/exec",
            json={"command": command, "timeout": timeout},
        )
        return ExecResult(**data)

    async def exec_stream(self, command: str) -> AsyncIterator[dict]:
        """Execute with streaming output via WebSocket."""
        ws_url = self._client.ws_url(f"/sandboxes/{self.id}/ws")
        extra_headers = {}
        if self._client.api_key:
            extra_headers["Authorization"] = f"Bearer {self._client.api_key}"

        async with websockets.connect(ws_url, additional_headers=extra_headers) as ws:
            # Wait for ready
            msg = json.loads(await ws.recv())
            if msg.get("type") != "ready":
                raise RuntimeError(f"Expected ready, got: {msg}")

            # Send exec command
            await ws.send(json.dumps({"type": "exec", "command": command}))

            # Stream responses
            async for raw in ws:
                msg = json.loads(raw)
                msg_type = msg.get("type")

                if msg_type == "stdout":
                    decoded = base64.b64decode(msg["data"]).decode("utf-8", errors="replace")
                    yield {"type": "stdout", "data": decoded}
                elif msg_type == "stderr":
                    decoded = base64.b64decode(msg["data"]).decode("utf-8", errors="replace")
                    yield {"type": "stderr", "data": decoded}
                elif msg_type == "exit":
                    yield {"type": "exit", "code": msg["code"]}
                    break
                elif msg_type == "error":
                    yield {"type": "error", "message": msg["message"]}
                    break

    # ── File operations ─────────────────────────────────────────

    def upload(self, local_path: str, remote_path: str) -> None:
        """Upload a file to the sandbox."""
        with open(local_path, "rb") as f:
            self._client.post(
                f"/sandboxes/{self.id}/files",
                files={"file": f},
                data={"path": remote_path},
            )

    def upload_content(self, content: bytes, remote_path: str) -> None:
        """Upload content directly to the sandbox."""
        self._client.post(
            f"/sandboxes/{self.id}/files",
            files={"file": ("upload", content)},
            data={"path": remote_path},
        )

    def download(self, remote_path: str) -> bytes:
        """Download a file from the sandbox."""
        return self._client.get_bytes(
            f"/sandboxes/{self.id}/files",
            params={"path": remote_path},
        )

    def list_files(self, path: str = "/workspace") -> list[FileEntry]:
        """List files in the sandbox."""
        data = self._client.get(
            f"/sandboxes/{self.id}/files",
            params={"list": "true", "path": path},
        )
        return [FileEntry(**f) for f in data]

    def delete_file(self, path: str) -> None:
        """Delete a file in the sandbox."""
        self._client.delete(
            f"/sandboxes/{self.id}/files",
            params={"path": path},
        )

    def mkdir(self, path: str) -> None:
        """Create a directory in the sandbox."""
        self._client.put(
            f"/sandboxes/{self.id}/files",
            params={"path": path},
        )

    # ── Signals ─────────────────────────────────────────────────

    def send_signal(self, signal: int) -> None:
        """Send a POSIX signal to the sandbox process."""
        self._client.post(
            f"/sandboxes/{self.id}/signal",
            json={"signal": signal},
        )

    # ── Port forwarding ────────────────────────────────────────

    def port_forward(self, guest_port: int) -> PortForwardInfo:
        """Create a port forward from host to guest."""
        data = self._client.post(
            f"/sandboxes/{self.id}/ports",
            json={"guest_port": guest_port},
        )
        return PortForwardInfo(**data)

    def list_port_forwards(self) -> list[PortForwardInfo]:
        """List active port forwards."""
        data = self._client.get(f"/sandboxes/{self.id}/ports")
        return [PortForwardInfo(**p) for p in data["ports"]]

    def remove_port_forward(self, guest_port: int) -> None:
        """Remove a port forward."""
        self._client.delete(f"/sandboxes/{self.id}/ports/{guest_port}")

    # ── Info & lifecycle ────────────────────────────────────────

    def info(self) -> SandboxInfo:
        """Get sandbox info."""
        data = self._client.get(f"/sandboxes/{self.id}")
        return SandboxInfo(**data)

    def destroy(self) -> None:
        """Destroy the sandbox and its VM."""
        self._client.delete(f"/sandboxes/{self.id}")
        self._client.close()

    def __enter__(self) -> Sandbox:
        return self

    def __exit__(self, exc_type, exc_val, exc_tb) -> None:
        try:
            self.destroy()
        except Exception:
            if exc_type is None:
                raise

    # === LLM Tool Definitions ===

    def tool_definitions(self, format: str = "openai") -> list[dict]:
        """Return tool schemas for LLM function calling.

        Args:
            format: "openai", "anthropic", or "generic"
        """
        from .tools import get_tool_definitions
        return get_tool_definitions(format)

    def handle_tool_call(self, tool_call: dict, *, raise_on_error: bool = True) -> dict:
        """Execute an LLM tool call against this sandbox.

        Works with both OpenAI and Anthropic tool call formats.

        Args:
            tool_call: The tool call dict from the LLM.
            raise_on_error: If True (default), raise exceptions on errors.
                If False, return {"error": ...} dicts (legacy behavior).
        """
        from .tools import handle_tool_call
        return handle_tool_call(self, tool_call, raise_on_error=raise_on_error)

File: sdks/typescript/src/sandbox.ts
import { AgentBoxClient } from "./client.js";
import { getToolDefinitions, handleToolCall } from "./tools.js";
import type {
  ExecResult,
  ExecStreamEvent,
  FileEntry,
  HealthStatus,
  PoolStatus,
  PortForwardInfo,
  SandboxConfig,
  SandboxInfo,
  ToolResult,
} from "./types.js";

/** Interactive exec session with stdin and signal support. */
export class ExecSession {
  private readonly ws: WebSocket;
  private readonly messages: ExecStreamEvent[] = [];
  private resolve: (() => void) | null = null;
  private done = false;

  /** @internal */
  constructor(ws: WebSocket) {
    this.ws = ws;

    ws.onmessage = (event) => {
      const msg = JSON.parse(event.data as string);
      const type = msg.type as string;

      if (type === "stdout" || type === "stderr") {
        this.messages.push({ type, data: atob(msg.data) });
      } else if (type === "exit") {
        this.messages.push({ type: "exit", code: msg.code });
        this.done = true;
      } else if (type === "error") {
        this.messages.push({ type: "error", message: msg.message });
        this.done = true;
      }

      if (this.resolve) {
        this.resolve();
        this.resolve = null;
      }
    };

    ws.onerror = () => {
      this.messages.push({
        type: "error",
        message: "WebSocket error occurred",
      });
      this.done = true;
      if (this.resolve) {
        this.resolve();
        this.resolve = null;
      }
    };

    ws.onclose = () => {
      this.done = true;
      if (this.resolve) {
        this.resolve();
        this.resolve = null;
      }
    };
  }

  /** Async iterator yielding exec events. */
  async *events(): AsyncGenerator<ExecStreamEvent> {
    try {
      while (true) {
        if (this.messages.length > 0) {
          const msg = this.messages.shift()!;
          yield msg;
          if (msg.type === "exit" || msg.type === "error") break;
        } else if (this.done) {
          break;
        } else {
          await new Promise<void>((r) => {
            this.resolve = r;
          });
        }
      }
    } finally {
      this.ws.close();
    }
  }

  /** Send stdin bytes to the running command. */
  sendStdin(data: Uint8Array): void {
    const encoded = btoa(
      Array.from(data)
        .map((b) => String.fromCharCode(b))
        .join(""),
    );
    this.ws.send(JSON.stringify({ type: "stdin", data: encoded }));
  }

  /** Send a POSIX signal to the running command. */
  sendSignal(signal: number): void {
    this.ws.send(JSON.stringify({ type: "signal", signal }));
  }

  /** Close the WebSocket connection. */
  close(): void {
    this.ws.close();
  }
}

/** A sandboxed environment for executing code. */
export class Sandbox {
  readonly id: string;
  private readonly client: AgentBoxClient;

  private constructor(id: string, client: AgentBoxClient) {
    this.id = id;
    this.client = client;
  }

  /** Create a new sandbox. Boots a microVM in <300ms. */
  static async create(
    options?: SandboxConfig & { url?: string; api_key?: string },
  ): Promise<Sandbox> {
    const client = new AgentBoxClient(options?.url, options?.api_key);
    const body: Record<string, unknown> = {
      memory_mb: options?.memory_mb ?? 2048,
      vcpus: options?.vcpus ?? 2,
      network: options?.network ?? false,
      timeout: options?.timeout ?? 3600,
    };
    if (options?.disk_size_mb) body.disk_size_mb = options.disk_size_mb;
    const data = (await client.post("/sandboxes", body)) as SandboxInfo;
    return new Sandbox(data.id, client);
  }

  // ── Static API methods ──────────────────────────────────────

  /** List all active sandboxes. */
  static async list(options?: {
    url?: string;
    api_key?: string;
  }): Promise<SandboxInfo[]> {
    const client = new AgentBoxClient(options?.url, options?.api_key);
    return (await client.listSandboxes()) as SandboxInfo[];
  }

  /** Get pool status. */
  static async poolStatus(options?: {
    url?: string;
    api_key?: string;
  }): Promise<PoolStatus> {
    const client = new AgentBoxClient(options?.url, options?.api_key);
    return (await client.poolStatus()) as PoolStatus;
  }

  /** Health check (public, no auth required). */
  static async health(options?: { url?: string }): Promise<HealthStatus> {
    const client = new AgentBoxClient(options?.url);
    return (await client.health()) as HealthStatus;
  }

  // ── Command execution ───────────────────────────────────────

  /** Execute a command and wait for completion. */
  async exec(command: string, timeout = 30): Promise<ExecResult> {
    return (await this.client.post(`/sandboxes/${this.id}/exec`, {
      command,
      timeout,
    })) as ExecResult;
  }

  /** Execute with streaming output via WebSocket. Returns an async iterator. */
  async *execStream(command: string): AsyncGenerator<ExecStreamEvent> {
    const wsUrl = this.client.wsUrl(`/sandboxes/${this.id}/ws`);
    const ws = new WebSocket(wsUrl);

    const messages: ExecStreamEvent[] = [];
    let resolve: (() => void) | null = null;
    let done = false;

    ws.onmessage = (event) => {
      const msg = JSON.parse(event.data as string);
      const type = msg.type as string;

      if (type === "ready") {
        ws.send(JSON.stringify({ type: "exec", command }));
        return;
      }

      if (type === "stdout" || type === "stderr") {
        const decoded = atob(msg.data);
        messages.push({ type, data: decoded });
      } else if (type === "exit") {
        messages.push({ type: "exit", code: msg.code });
        done = true;
      } else if (type === "error") {
        messages.push({ type: "error", message: msg.message });
        done = true;
      }

      if (resolve) {
        resolve();
        resolve = null;
      }
    };

    ws.onerror = () => {
      messages.push({
        type: "error",
        message: "WebSocket error occurred",
      });
      done = true;
      if (resolve) {
        resolve();
        resolve = null;
      }
    };

    ws.onclose = () => {
      done = true;
      if (resolve) {
        resolve();
        resolve = null;
      }
    };

    // Wait for WebSocket to open
    await new Promise<void>((res, rej) => {
      ws.onopen = () => res();
      ws.onerror = () => rej(new Error("WebSocket connection failed"));
    });

    // Restore the streaming error handler (onopen promise overwrites it)
    ws.onerror = () => {
      messages.push({
        type: "error",
        message: "WebSocket error occurred",
      });
      done = true;
      if (resolve) {
        resolve();
        resolve = null;
      }
    };

    try {
      while (true) {
        if (messages.length > 0) {
          const msg = messages.shift()!;
          yield msg;
          if (msg.type === "exit" || msg.type === "error") break;
        } else if (done) {
          break;
        } else {
          await new Promise<void>((r) => {
            resolve = r;
          });
        }
      }
    } finally {
      ws.close();
    }
  }

  /** Start an interactive exec session with stdin and signal support. */
  async execInteractive(command: string): Promise<ExecSession> {
    const wsUrl = this.client.wsUrl(`/sandboxes/${this.id}/ws`);
    const ws = new WebSocket(wsUrl);

    // Wait for open
    await new Promise<void>((res, rej) => {
      ws.onopen = () => res();
      ws.onerror = () => rej(new Error("WebSocket connection failed"));
    });

    // Wait for ready
    await new Promise<void>((res, rej) => {
      ws.onmessage = (event) => {
        const msg = JSON.parse(event.data as string);
        if (msg.type === "ready") {
          res();
        } else {
          rej(new Error(`Expected ready, got: ${JSON.stringify(msg)}`));
        }
      };
      ws.onerror = () =>
        rej(new Error("WebSocket error while waiting for ready"));
    });

    // Send exec command
    ws.send(JSON.stringify({ type: "exec", command }));

    return new ExecSession(ws);
  }

  // ── File operations ─────────────────────────────────────────

  /** Upload content to the sandbox. */
  async uploadContent(content: Uint8Array, remotePath: string): Promise<void> {
    const form = new FormData();
    form.append("path", remotePath);
    form.append("file", new Blob([content]), "upload");
    await this.client.postMultipart(`/sandboxes/${this.id}/files`, form);
  }

  /** Download a file from the sandbox. */
  async download(remotePath: string): Promise<Uint8Array> {
    const buf = await this.client.getBytes(`/sandboxes/${this.id}/files`, {
      path: remotePath,
    });
    return new Uint8Array(buf);
  }

  /** List files in the sandbox. */
  async listFiles(path = "/workspace"): Promise<FileEntry[]> {
    return (await this.client.get(`/sandboxes/${this.id}/files`, {
      list: "true",
      path,
    })) as FileEntry[];
  }

  /** Delete a file in the sandbox. */
  async deleteFile(path: string): Promise<void> {
    await this.client.delete(`/sandboxes/${this.id}/files`, { path });
  }

  /** Create a directory in the sandbox. */
  async mkdir(path: string): Promise<void> {
    await this.client.put(`/sandboxes/${this.id}/files`, { path });
  }

  // ── Signals ─────────────────────────────────────────────────

  /** Send a POSIX signal to the sandbox process. */
  async sendSignal(signal: number): Promise<void> {
    await this.client.post(`/sandboxes/${this.id}/signal`, { signal });
  }

  // ── Port forwarding ────────────────────────────────────────

  /** Forward a guest port to a host port. Returns the allocated host address. */
  async portForward(guestPort: number): Promise<PortForwardInfo> {
    return (await this.client.post(`/sandboxes/${this.id}/ports`, {
      guest_port: guestPort,
    })) as PortForwardInfo;
  }

  /** List active port forwards for this sandbox. */
  async listPortForwards(): Promise<PortForwardInfo[]> {
    const data = (await this.client.get(
      `/sandboxes/${this.id}/ports`,
    )) as { ports: PortForwardInfo[] };
    return data.ports;
  }

  /** Remove a port forward by guest port. */
  async removePortForward(guestPort: number): Promise<void> {
    await this.client.delete(`/sandboxes/${this.id}/ports/${guestPort}`);
  }

  // ── Info & lifecycle ────────────────────────────────────────

  /** Get sandbox info. */
  async info(): Promise<SandboxInfo> {
    return (await this.client.get(`/sandboxes/${this.id}`)) as SandboxInfo;
  }

  /** Destroy the sandbox and its VM. */
  async destroy(): Promise<void> {
    await this.client.delete(`/sandboxes/${this.id}`);
  }

  // ── LLM Tools ───────────────────────────────────────────────

  /** Return tool schemas for LLM function calling. */
  toolDefinitions(
    format: "openai" | "anthropic" | "generic" = "openai",
  ): unknown[] {
    return getToolDefinitions(format);
  }

  /** Execute an LLM tool call against this sandbox. */
  async handleToolCall(
    toolCall: Record<string, unknown>,
  ): Promise<ToolResult> {
    return handleToolCall(this, toolCall);
  }
}
